### ISSUE TITLE
Default service account on Compute Engine

### ISSUE THREAD

**Issue by nkovacs** [kreait/firebase-php#229]
I would like to use this library on Google's Compute Engine with the vm's default service account, without using a credentials json file.

Replacing ServiceAccountCredentials at https://github.com/kreait/firebase-php/blob/master/src/Firebase/Factory.php#L306 with [GCECredentials](https://github.com/google/google-auth-library-php/blob/master/src/Credentials/GCECredentials.php) works, but the library still needs a ServiceAccount to get the project id and the private key for signing custom tokens.

The project id can also be retrieved from the metadata service that GCECredentials uses: https://cloud.google.com/compute/docs/storing-retrieving-metadata

Signing custom tokens should be done using the iam signblob api call: https://firebase.google.com/docs/auth/admin/create-custom-tokens#using_a_service_account_id

**Comment by jeromegamez** [comment]
Thanks for bringing this up!

Concerning the custom token signing, this is a relatively new feature (I only knew about it from the official SDK announcements) and I will implement it in an upcoming release.

I'll also refactor the credentials - they will probably be more tightly coupled to guzzle which I wanted to avoid, but that's for the next major release anyways.

In the meantime, if you haven't already, could you please try initializing the project without a service account at all? The SDK will then try to autodiscover credentials - I'm not too optimistic this will work, but it's worth a shot :)

```php
use Kreait\Firebase\Factory;

$firebase = (new Factory)->create();
```

**Comment by jeromegamez** [comment]
After looking a little bit into it, signing custom tokens with just the service account ID has one drawback: it will result in an additional request to the IAM Api, the same applies for fetching the project id.

This doesn't mean that it shouldn't be implemented, but I think it's worth noticing.

**Comment by nkovacs** [comment]
Hi

No, omitting the service account just causes the default discoverers to run, which fail to find a service account json, so the factory throws an error.

GCECredentials also has to do a request to the metadata endpoint. It does support some kind of cache, but I don't know how that works and I can't find any documentation about it. It looks like that package is not meant to be used directly, but through the google/apiclient package, which supports caching: https://github.com/google/google-api-php-client#caching

**Comment by jeromegamez** [comment]
This will require some core changes in the SDK (factory) and I have to ensure that nothing breaks/works as before for existing users of the SDK, so I can not promise that a solution for this will be available soon (nor that it will be pretty 😭).

@nkovacs As I don't have a GCE instance available for testing, can I trigger you as soon as I publish a WIP-Pull Request to tackle this so that you'll have a look if everything works as expected?

**Comment by jeromegamez** [comment]
@nkovacs Could you please test if #230 works for you?

**Comment by nkovacs** [comment]
Thanks for doing this so quickly.

I'm getting this error:

```
Exception 'TypeError' with message 'Return value of Kreait\Firebase\ServiceAccount::getClientId() must be of the type string, null returned'

in /var/www/app/releases/20180730105743/vendor/kreait/firebase-php/src/Firebase/ServiceAccount.php:43

Stack trace:
#0 /var/www/app/releases/20180730105743/vendor/kreait/firebase-php/src/Firebase/Factory.php(351): Kreait\Firebase\ServiceAccount->getClientId()
#1 /var/www/app/releases/20180730105743/vendor/kreait/firebase-php/src/Firebase/Factory.php(336): Kreait\Firebase\Factory->getGoogleCloudServiceBuilder()
#2 /var/www/app/releases/20180730105743/vendor/kreait/firebase-php/src/Firebase/Factory.php(158): Kreait\Firebase\Factory->createStorage()
#3 /var/www/app/releases/20180730105743/components/firebase/Component.php(27): Kreait\Firebase\Factory->create()
```

The code is just `(new Factory())->create()->getMessaging()`.

I've changed getGoogleCloudServiceBuilder to this:

```
    protected function getGoogleCloudServiceBuilder(): ServiceBuilder
    {
        $serviceAccount = $this->getServiceAccount();

        if ((new GcpMetadata())->isAvailable()) {
            $options = [];
        } else {
            $options = [
                'keyFile' => [
                    'client_email' => $serviceAccount->getClientEmail(),
                    'client_id' => $serviceAccount->getClientId(),
                    'private_key' => $serviceAccount->getPrivateKey(),
                    'type' => 'service_account',
                ],
            ];
        }

        return new ServiceBuilder($options);
    }
```

I've only tested push notifications, and those work after this change. I'll test the rest, but we're only using push notifications, so I'll have to write some code.

**Comment by jeromegamez** [comment]
@nkovacs I just updated the PR, could you please try again? And could we please move the conversation over to #230 to have it more close to the code 😅?

**Comment by jeromegamez** [comment]
Released with [4.15.0](https://github.com/kreait/firebase-php/releases/tag/4.15.0) 🎉
